mmedit/apis/inferencers/base_mmedit_inferencer.py

```diff
@@ -0,0 +1,210 @@
+        device (str, optional): Device to run inference. If None, the best
+            device will be automatically used.
+        result_out_dir (str): Output directory of images. Defaults to ''.
+    """
+
+    func_kwargs = dict(
+        preprocess=[],
+        forward=[],
+        visualize=['result_out_dir'],
+        postprocess=['get_datasample'])
+    func_order = dict(preprocess=0, forward=1, visualize=2, postprocess=3)
+
+    extra_parameters = dict()
+
+    def __init__(self,
+                 config: Union[ConfigType, str],
+                 ckpt: Optional[str],
+                 device: Optional[str] = None,
+                 extra_parameters: Optional[Dict] = None,
+                 **kwargs) -> None:
+        # Load config to cfg
+        if isinstance(config, str):
+            cfg = Config.fromfile(config)
+        elif not isinstance(config, ConfigType):
+            raise TypeError('config must be a filename or any ConfigType'
+                            f'object, but got {type(cfg)}')
+        self.cfg = cfg
+        if cfg.model.get('pretrained'):
+            cfg.model.pretrained = None
+
+        if device is None:
+            device = torch.device(
+                'cuda' if torch.cuda.is_available() else 'cpu')
+        self.device = device
+        self._init_model(cfg, ckpt, device)
+        self._init_extra_parameters(extra_parameters)
+        self.base_params = self._dispatch_kwargs(**kwargs)
+
+    def _init_model(self, cfg: Union[ConfigType, str], ckpt: Optional[str],
+                    device: str) -> None:
+        """Initialize the model with the given config and checkpoint on the
+        specific device."""
+        model = MODELS.build(cfg.model)
+        if ckpt is not None and ckpt != '':
+            ckpt = load_checkpoint(model, ckpt, map_location='cpu')
+        model.cfg = cfg
+        model.to(device)
+        model.eval()
+        self.model = model
+
+    def _init_extra_parameters(self, extra_parameters: Dict) -> None:
+        """Initialize extra_parameters of each kind of inferencer."""
+        if extra_parameters is not None:
+            for key in self.extra_parameters.keys():
+                if key in extra_parameters.keys():
+                    self.extra_parameters[key] = extra_parameters[key]
+
+    def _dispatch_kwargs(self, **kwargs) -> Tuple[Dict, Dict, Dict, Dict]:
+        """Dispatch kwargs to preprocess(), forward(), visualize() and
+        postprocess() according to the actual demands."""
+        results = [{}, {}, {}, {}]
+        dispatched_kwargs = set()
+
+        # Dispatch kwargs according to self.func_kwargs
+        for func_name, func_kwargs in self.func_kwargs.items():
+            for func_kwarg in func_kwargs:
+                if func_kwarg in kwargs:
+                    dispatched_kwargs.add(func_kwarg)
+                    results[self.func_order[func_name]][func_kwarg] = kwargs[
+                        func_kwarg]
+
+        return results
+
```

Comment:
How about packing `__call__` with `@torch.no_grad` to save GPU memory?

Reply:
good idea. resolved.